### PR TITLE
Enable reognization on partition leaf node

### PR DESCRIPTION
**Pull request by baobao0206**
We are not allowed to modify the distribution of partition leaf table
after gpdb6. Because the distibution of root table and leaf table must
be same. But if only reorg on the partition leaf table, it will not
cause this problem. So we enable reorg when the distibution was not be
changed.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/sql/alter_distribution_policy.sql [hunk 1]
```diff
@@ -428,3 +428,12 @@ CREATE TABLE alter_table_with_primary_key (a int primary key);
 ALTER TABLE alter_table_with_primary_key SET DISTRIBUTED RANDOMLY;
 CREATE TABLE alter_table_with_unique_index (a int unique);
 ALTER TABLE alter_table_with_unique_index SET DISTRIBUTED RANDOMLY;
+
+-- Enable reorg partition leaf table
+create table reorg_leaf (trans_id int, date text, amount decimal(9,2), region text) distributed by (trans_id)
+partition by range(date)
+(partition p2011 start (date '2011-01-01'::text) end (date '2012-01-01'::text),
+	partition p2012 start (date '2012-01-01'::text) end (date '2013-01-01'::text));
+alter table reorg_leaf_1_prt_p2011 set with (reorganize=true) distributed by (trans_id);
```
Comment (ashwinstar) [inline src/test/regress/sql/alter_distribution_policy.sql:437]: I feel we should add some minimal data as well. Plus, also validate the distribution policy didn't change after reorg in test. Just checking no error was produced seems not enough.

File: src/test/regress/sql/alter_distribution_policy.sql [hunk 2]
```diff
@@ -428,3 +428,12 @@ CREATE TABLE alter_table_with_primary_key (a int primary key);
...
+alter table reorg_leaf_1_prt_p2011 set with (reorganize=true);
```
Comment (ashwinstar) [inline src/test/regress/sql/alter_distribution_policy.sql:438]: Do we have test to validate negative case, where reorg should fail if distribution policy is not same as before?

File: src/backend/commands/tablecmds.c [hunk 3]
```diff
@@ -15404,6 +15471,8 @@ ATExecSetDistributedBy(Relation rel, Node *node, AlterTableCmd *cmd)
 			(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
 			errmsg("permission denied: \"%s\" is a system catalog", RelationGetRelationName(rel))));
 
+
```
[the remark this answers is not in the document]
Comment (baobao0206) [reply src/backend/commands/tablecmds.c:15474]: It is a mistake.

File: src/test/regress/sql/alter_distribution_policy.sql (hunk 2 above)
Comment (baobao0206) [reply src/test/regress/sql/alter_distribution_policy.sql]: Have add the test case to validate negative case

File: src/test/regress/sql/alter_distribution_policy.sql (hunk 1 above)
Comment (baobao0206) [reply src/test/regress/sql/alter_distribution_policy.sql]: Add some data

File: src/test/regress/sql/alter_distribution_policy.sql [hunk 4]
```diff
@@ -428,3 +428,12 @@ CREATE TABLE alter_table_with_primary_key (a int primary key);
 ALTER TABLE alter_table_with_primary_key SET DISTRIBUTED RANDOMLY;
 CREATE TABLE alter_table_with_unique_index (a int unique);
 ALTER TABLE alter_table_with_unique_index SET DISTRIBUTED RANDOMLY;
+
+-- Enable reorg partition leaf table
+create table reorg_leaf (trans_id int, date text, amount decimal(9,2), region text) distributed by (trans_id)
```
[the remark this answers is not in the document]
Comment (baobao0206) [reply src/test/regress/sql/alter_distribution_policy.sql]: Have update the test case.

File: src/backend/commands/tablecmds.c [hunk 5]
```diff
@@ -4541,6 +4548,66 @@ ATPrepCmd(List **wqueue, Relation rel, AlterTableCmd *cmd,
...
 						case PART_STATUS_LEAF:
+							lprime = (List *)cmd->def;
+							Assert(PointerIsValid(cmd->def));
+							Assert(IsA(cmd->def, List));
+							ldistro = (DistributedBy *)lsecond(lprime);
+							if (ldistro == NULL)
+								break;
+							ldistro->numsegments = rel->rd_cdbpolicy->numsegments;
+
+							foreach(lc, ldistro->keyCols)
```
[the remark this answers is not in the document]
Comment (baobao0206) [reply src/backend/commands/tablecmds.c]: Yes, use getPolicyForDistributedBy is better. Have rewrite the code.

File: src/test/regress/sql/alter_distribution_policy.sql [hunk 6]
```diff
@@ -428,3 +428,17 @@ CREATE TABLE alter_table_with_primary_key (a int primary key);
 ALTER TABLE alter_table_with_primary_key SET DISTRIBUTED RANDOMLY;
 CREATE TABLE alter_table_with_unique_index (a int unique);
 ALTER TABLE alter_table_with_unique_index SET DISTRIBUTED RANDOMLY;
+
+-- Enable reorg partition leaf table
+create table reorg_leaf (a int, b int, c int) distributed by (c)
```
Comment (ashwinstar) [inline src/test/regress/sql/alter_distribution_policy.sql:433]: I hope you tried sub-partitions and even for middle level parent partitions, modifying the distribution continues to be prohibited.

Comment (ashwinstar) [inline src/test/regress/sql/alter_distribution_policy.sql:433]: Do we have test where top level parent partition distribution key can be changed and it modifies for full hierarchy?

File: src/test/regress/sql/alter_distribution_policy.sql [hunk 7]
```diff
@@ -428,3 +428,17 @@ CREATE TABLE alter_table_with_primary_key (a int primary key);
...
+partition by range(a)
+(partition p0 start (0) end (5),
+	partition p1 start (5) end (10));
+insert into reorg_leaf select i % 20, i, i from generate_series(0, 9) i;
```
Comment (ashwinstar) [inline src/test/regress/sql/alter_distribution_policy.sql:437]: What's the point of doing `i%20` when we are inserting only from 0-9?

File: src/test/regress/sql/alter_distribution_policy.sql (hunk 6 above)
Comment (baobao0206) [reply src/test/regress/sql/alter_distribution_policy.sql:433]: Yes, we have test it on some cases above.

File: src/test/regress/sql/alter_distribution_policy.sql (hunk 7 above)
Comment (baobao0206) [reply src/test/regress/sql/alter_distribution_policy.sql:437]: 10% is better.

File: src/test/regress/sql/alter_distribution_policy.sql (hunk 6 above)
Comment (baobao0206) [reply src/test/regress/sql/alter_distribution_policy.sql:433]: Have add check on middle level.

**Comment by baobao0206** [comment]
@ashwinstar Thanks for your reviewing. I have update the code and test cases depending on the comments.

**Comment by ashwinstar** [comment]
> On May 17, 2020, at 12:34 AM, Jinbao Chen <notifications@github.com> wrote:

> @ashwinstar Thanks for your reviewing. I have update the code and test cases depending on the comments.

Looks good to me. Shit it!

**Pull request merged**